packages/protocol/contracts/bridge/libs/LibBridgeRelease.sol

```diff
@@ -56,37 +66,73 @@ library LibBridgeRelease {
 
-        if (state.etherReleased[msgHash] == true) {
-            revert B_ETHER_RELEASED_ALREADY();
-        }
-
         if (
             !LibBridgeStatus.isMessageFailed(
                 resolver, msgHash, message.destChainId, proof
             )
         ) {
             revert B_MSG_NOT_FAILED();
         }
+        
+        if(state.recallStatus[msgHash] 
+                == LibBridgeData.RecallStatus.FULLY_RECALLED
+        ){
+            // Both ether and tokens are released
+            revert B_TOKENS_RELEASED_ALREADY();
+        }
 
-        state.etherReleased[msgHash] = true;
+        uint256 releaseAmount;
 
-        uint256 releaseAmount = message.depositValue + message.callValue;
+        if(state.recallStatus[msgHash] 
+                == LibBridgeData.RecallStatus.NOT_RECALLED
+        ) {
+            // Release ETH first
+            state.recallStatus[msgHash] = LibBridgeData.RecallStatus.ETH_RELEASED;
 
-        if (releaseAmount > 0) {
-            address ethVault = resolver.resolve("ether_vault", true);
-            // if on Taiko
-            if (ethVault != address(0)) {
-                EtherVault(payable(ethVault)).releaseEther(
-                    message.owner, releaseAmount
-                );
-            } else {
-                // if on Ethereum
-                (bool success,) = message.owner.call{ value: releaseAmount }("");
-                if (!success) {
-                    revert B_FAILED_TRANSFER();
+            releaseAmount = message.depositValue + message.callValue;
+
+            if (releaseAmount > 0) {
+                address ethVault = resolver.resolve("ether_vault", true);
+                // if on Taiko
+                if (ethVault != address(0)) {
+                    EtherVault(payable(ethVault)).releaseEther(
+                        message.owner, releaseAmount
+                    );
+                } else {
+                    // if on Ethereum
+                    (bool success,) = message.owner.call{ value: releaseAmount }("");
+                    if (!success) {
+                        revert B_FAILED_TRANSFER();
+                    }
                 }
             }
         }
+        //2nd stage is releasing the tokens
+        if(state.recallStatus[msgHash] 
+                == LibBridgeData.RecallStatus.ETH_RELEASED 
+        ) {
+            if( message.to == address(0)
+                || !_isContract(message.sender)
+                || !IERC165Upgradeable(message.sender).supportsInterface(VAULT_INTERFACE_ID) 
+            ) {
+                state.recallStatus[msgHash] = LibBridgeData.RecallStatus.FULLY_RECALLED;
+            } else {
+                try VaultContract(
+                    (message.sender)
+                ).releaseToken(message){
+                    state.recallStatus[msgHash] =
+                        LibBridgeData.RecallStatus.FULLY_RECALLED;
+                } catch {}
+            }
+        }
```

Comment:
rename EtherReleased to "MessageRecalled", and add one extra field for value: state.recallStatus[msgHash]`.

I also think we should use `message.sender`, not `message.owner` in this event.